docs/decisions/0011-kernel-service-registration.md

```diff
@@ -0,0 +1,180 @@
+
+```csharp
+public TextMemoryPlugin(ISemanticTextMemory memory)
+{
+    this._memory = memory;
+}
+```
+
+`TextMemoryPlugin` depends on `ISemanticTextMemory` interface. In similar way, other Plugins may have multiple dependencies and there should be a way how to resolve required dependencies manually or automatically.
+
+At the moment, `ISemanticTextMemory` is a property of `IKernel` interface, which allows to inject `ISemanticTextMemory` into `TextMemoryPlugin` during Plugin initialization:
+
+```csharp
+kernel.ImportFunctions(new TextMemoryPlugin(kernel.Memory));
+```
+
+There should be a way how to support not only Memory-related interface, but any kind of service, which can be used in Plugin - `ISemanticTextMemory`, `IPromptTemplateEngine`, `IDelegatingHandlerFactory` or any other service.
+
+## Considered Options
+
+### Solution #1.1 (available by default)
+
+User is responsible for all Plugins initialization and dependency resolution with **manual** approach.
+
+```csharp
+var memoryStore = new VolatileMemoryStore();
+var embeddingGeneration = new OpenAITextEmbeddingGeneration(modelId, apiKey);
+var semanticTextMemory = new SemanticTextMemory(memoryStore, embeddingGeneration);
+
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+var kernel = Kernel.Builder.Build();
+
+kernel.ImportFunctions(memoryPlugin);
+```
+
+Note: this is native .NET approach how to resolve service dependencies manually, and this approach should always be available by default. Any other solutions which could help to improve dependency resolution can be added on top of this approach.
+
+### Solution #1.2 (available by default)
+
+User is responsible for all Plugins initialization and dependency resolution with **dependency injection** approach.
+
+```csharp
+var serviceCollection = new ServiceCollection();
+
+serviceCollection.AddTransient<IMemoryStore, VolatileMemoryStore>();
+serviceCollection.AddTransient<ITextEmbeddingGeneration>(
+    (serviceProvider) => new OpenAITextEmbeddingGeneration(modelId, apiKey));
+
+serviceCollection.AddTransient<ISemanticTextMemory, SemanticTextMemory>();
+
+var services = serviceCollection.BuildServiceProvider();
+
+// In theory, TextMemoryPlugin can be also registered in DI container.
+var memoryPlugin = new TextMemoryPlugin(services.GetService<ISemanticTextMemory>());
+
+var kernel = Kernel.Builder.Build();
+
+kernel.ImportFunctions(memoryPlugin);
+```
+
+Note: in similar way as Solution #1.1, this way should be supported out of the box. Users always can handle all the dependencies on their side and just provide required Plugins to Kernel.
+
+### Solution #2.1
+
+Custom service collection and service provider on Kernel level to simplify dependency resolution process, as addition to Solution #1.1 and Solution #1.2.
```

Comment:
In the fullness of time, isn't it likely we're just going to find ourselves implementing a full-blown DI container, as more and more requests come in for this or that feature that needs to be supported?